### ISSUE TITLE
Accessibility: Make app navigation with dismissable drawers accessible for screen readers

### ISSUE THREAD

**Issue by maracuja-juice** [material-components/material-components-web#4704]
I want to create an accessible navigation using Material Components Web.
The basic html markup that would be something like this:
```
<nav id="navigation">
  <button aria-expanded="false">Menu</button>
  <ul hidden>
    <li><a href="/">Home</a></li>
    <li><a href="/about">About</a></li>
    <li><a href="/shop">Shop</a></li>
    <li><a href="/contact">Contact</a></li>
  </ul>
</nav>
```
(This is taken from [this great resource](https://inclusive-components.design/menus-menu-buttons))
It's important that the menu button is **right next** to the ul, otherwise screenreaders won't know that the button controls the state of the ul.

Using this kind of markup doesn't seem to be supported right now as in the examples you use a different markup which **isn't** accessible. I'm referring to [this](https://material.io/develop/web/components/drawers/#dismissible-drawer-below-top-app-bar).

It should be possible in MDC to have the list of navigation items right next to menu button. How the exact markup for that should look like, I'm not sure yet myself.

If this is actually possible and I just wasn't able to achieve it yet, then I suggest that the docs only show this accessible approach. 

Resources that I gathered for an accessible Menu:
- https://vimeo.com/album/5937296/video/331530115 (In the video it is suggested that aria-controls would work well for this markup but [it isn't actually the case](https://inclusive-components.design/menus-menu-buttons/#ariacontrols) The other points hold true though.)
Timestamp where she starts to talk about the menu: 9:46 
(The other parts are very valuable as well though if you want to learn some more stuff about accessibility.)
- https://inclusive-components.design/menus-menu-buttons/

**Comment by touficbatache** [comment]
Hi @maracuja-juice, I think the component you are looking for is the [Menu component](https://material.io/develop/web/components/menus/). If you take a closer look to this docs page, you'll find a [section](https://material.io/develop/web/components/menus/#anchor-to-element-within-wrapper) for anchoring and opening a menu from a button. If you're still not sure if this is the right component, make sure to check out the [demo page](https://material-components.github.io/material-components-web-catalog/#/component/menu) for this component.

**Comment by maracuja-juice** [comment]
From material.io: `Menus display a list of choices on temporary surfaces.`
And for Drawers: `Navigation drawers provide access to destinations in your app.` 
Menu's are more for dropdown Menus. So no I don't think that the Menu component is the right component for this as I want to make a navigation.
There is no example for a Navigation with a Menu on material.io whereas there are many with the Drawer.

**Comment by abhiomkar** [comment]
Menu component can also be used for navigation.

**Comment by maracuja-juice** [comment]
Okay. I will try that of course.
But with the drawer this is still an a11y issue.

**Comment by abhiomkar** [comment]
Can you please provide reproducible steps for Drawer's a11y issue?

**Comment by maracuja-juice** [comment]
Not sure what you need exactly. The docs already show the markup that's needed.  
-> https://material.io/develop/web/components/drawers/#dismissible-drawer-below-top-app-bar

I am not that familiar with using a screenreader myself but as the linked sources show (Sara Soueidan, Inclusive Components) this is definitely not detectable with a screenreader and the current markup.

**Comment by maracuja-juice** [comment]
btw, not sure I understood that correctly: Are you suggesting (@touficbatache) that I can somehow use the Menu component as a drawer?
Because I don't want to use a Menu for Navigation but an (accessible) Drawer.

**Comment by abhiomkar** [comment]
It would be helpful if you can share your existing demo / example or mockup that you're trying to accomplish.

About the original issue that you posted, I suggest you to use MDCMenu with anchor elements as menu/list items.

**Comment by maracuja-juice** [comment]
1: I thought the code sample from your docs should have been enough but here's another one: https://codesandbox.io/s/oj0m760kpq 
I want this kind of drawer but in an accessible way.
2: I am still not convinced that Menu is the right component for what I want because it is for `a temporary surface.` It wasn't made for drawers, that's what the drawer component was made for.

**Comment by touficbatache** [comment]
@maracuja-juice Can you please show us in an illustration what do you exactly want to achieve? I still haven't understood what you're looking for. Maybe link us to some websites you'd like your menu to look like?

**Comment by maracuja-juice** [comment]
@touficbatache 
Screenshot from Codesandbox:
<img src="https://user-images.githubusercontent.com/16801528/57882796-d9381d80-7824-11e9-97f2-57fadd9d6c8a.jpeg" width="400" />

Or something like this (screenshot from material.io docs):
<img src="https://storage.googleapis.com/spec-host-backup/mio-design%2Fassets%2F1JkhpaOoDpFrCKLLoZMmbfiBMofzKS-lY%2Fusage-do.png" width="400" />

Or as a last example from the material components web catalog:
<img src="https://user-images.githubusercontent.com/16801528/57883053-71ce9d80-7825-11e9-96cd-1bb3b69c4ea2.jpeg" width="400" />

**Comment by touficbatache** [comment]
@maracuja-juice What's the problem exactly? Are you not able to reproduce the drawable shown in the images?

**Comment by maracuja-juice** [comment]
The look is not the problem but the accessibility of the component: The markup in that last example is not completely accessible for screen readers. I outlined that in more detail in the initial description.

**Comment by moog16** [comment]
What is the description of the a11y issue? What is the expected announcement, and what is actually happening?

**Comment by maracuja-juice** [comment]
A screenreader can currently not understand the relationship between the button and the menu. And therefore can't tell you what the button does. For that the menu has to be a direct sibling.

I would really recommend to read/watch the ressources I posted above so you can maybe understand it a bit better.

**Comment by abhiomkar** [comment]
MDC Web does not provide Menu Button component yet. We suggest implementing Menu Button on your own by attaching it to the MDC Menu. The aria tags including `aria-expanded="false"` can be updated based on Menu open / close by listening to MDC Menus' lifecycle events. Please see [mdc-menu/README.md](https://github.com/material-components/material-components-web/tree/master/packages/mdc-menu) for more details.

This seems like a duplicate of #4072. I'm closing this one.

**Comment by maracuja-juice** [comment]
I don't think that this is a duplicate because this **isn't** about the menu component!
So I kindly suggest to reopen this issue. Maybe we should jump on Discord? Because I don't think the problem was understood yet

The recommended markup in the documentation is: (https://material-components.github.io/material-components-web-catalog/#/component/drawer)
```
<body>
  <aside class="mdc-drawer mdc-drawer--dismissible">
    <div class="mdc-drawer__content">
      <div class="mdc-list">
        <a class="mdc-list-item mdc-list-item--activated" href="#" aria-current="page">
          <i class="material-icons mdc-list-item__graphic" aria-hidden="true">inbox</i>
          <span class="mdc-list-item__text">Inbox</span>
        </a>
        <a class="mdc-list-item" href="#">
          <i class="material-icons mdc-list-item__graphic" aria-hidden="true">send</i>
          <span class="mdc-list-item__text">Outgoing</span>
        </a>
        <a class="mdc-list-item" href="#">
          <i class="material-icons mdc-list-item__graphic" aria-hidden="true">drafts</i>
          <span class="mdc-list-item__text">Drafts</span>
        </a>
      </div>
    </div>
  </aside>

  <div class="mdc-drawer-app-content">
    <header class="mdc-top-app-bar app-bar" id="app-bar">
      <div class="mdc-top-app-bar__row">
        <section class="mdc-top-app-bar__section mdc-top-app-bar__section--align-start">
          <a href="#" class="demo-menu material-icons mdc-top-app-bar__navigation-icon">menu</a>
          <span class="mdc-top-app-bar__title">Dismissible Drawer</span>
        </section>
      </div>
    </header>

    <main class="main-content" id="main-content">
      <div class="mdc-top-app-bar--fixed-adjust">
        App Content
      </div>
    </main>
  </div>
</body>
```

But it should be more like this:
```
<body>
  <div class="mdc-drawer-app-content">
    <header class="mdc-top-app-bar app-bar" id="app-bar">
      <div class="mdc-top-app-bar__row">
        <section class="mdc-top-app-bar__section mdc-top-app-bar__section--align-start">
          <button class="demo-menu material-icons mdc-top-app-bar__navigation-icon">menu</button>
		  <div class="mdc-list">
	        <a class="mdc-list-item mdc-list-item--activated" href="#" aria-current="page">
	          <i class="material-icons mdc-list-item__graphic" aria-hidden="true">inbox</i>
	          <span class="mdc-list-item__text">Inbox</span>
	        </a>
	        <a class="mdc-list-item" href="#">
	          <i class="material-icons mdc-list-item__graphic" aria-hidden="true">send</i>
	          <span class="mdc-list-item__text">Outgoing</span>
	        </a>
	        <a class="mdc-list-item" href="#">
	          <i class="material-icons mdc-list-item__graphic" aria-hidden="true">drafts</i>
	          <span class="mdc-list-item__text">Drafts</span>
	        </a>
	      </div>
          <span class="mdc-top-app-bar__title">Dismissible Drawer</span>

        </section>
      </div>
    </header>

    <main class="main-content" id="main-content">
      <div class="mdc-top-app-bar--fixed-adjust">
        App Content
      </div>
    </main>
  </div>
</body>
```
(maybe with li/ul tags in the mdc-list div)

And yes the aria-expanded is easy to implement. But that the navigation list is a direct sibling and still styles correctly is not.
Both things are very important for the screen reader to recognize this relationship correctly though.

**Comment by abhiomkar** [comment]
Thanks for sharing the markup. As per your example it is technically not a Menu component.

What you're suggesting is a different structure of markup for both Drawer & Top App Bar where Drawer items are composed within Top App Bar. I wouldn't keep the Navigation Drawer in the scope of Top App Bar instead both of these components can be controlled by parent component.

**Comment by maracuja-juice** [comment]
Yes.
I'm suggesting that the items that are in the drawer should be adjacent to the button (not a link as in the material examples) that controls the drawer. This way the accessiblity issue will be gone.

**Comment by abhiomkar** [comment]
That totally depends on specific app use case. For example consider below layout:

![image](https://user-images.githubusercontent.com/47312/58487722-18744180-8136-11e9-886c-eb8575d9ab3b.png)

Top App Bar's content includes Drawer and main app content, but in other use case, the drawer is rendered beside the Top App Bar:

![image](https://user-images.githubusercontent.com/47312/58487951-80c32300-8136-11e9-9cd6-2bf3fa34e68a.png)

The Icon button that toggles the drawer is always owned by Top App Bar, but neither of the component controls other component to support this case.

This accessibility feature to toggle `aria-expanded` needs to be handled by clients.

**Comment by maracuja-juice** [comment]
On handling aria-expanded we already agree on.

Let me reiterate that it is important for this accessiblity issue to be resolved that the menu items have to be adjacent to the button that controls it in the HTML document order. 

It can be rendered beside the top app bar or below, doesn't matter much where. For the screen reader it only matters that it knows what content the button controls and the only way for that to happen correctly is to have it adjacent. I think this concept still wasn't understood clearly yet. 
(If you don't believe me, check out the sources I posted)

To summarize:
- It's not important where it is rendered but the order in the HTML document is what's important for screen readers and that's not achievable + documented right now with the library.

I hope we can finally agree that:
1: This is an accessiblity issue
2: This is the way to fix it

**Comment by abhiomkar** [comment]
Are you referring to this example [Dismissible Drawer Below Top App Bar](https://github.com/material-components/material-components-web/tree/master/packages/mdc-drawer#dismissible-drawer-below-top-app-bar)?

DOM structure looks correct to me.

**Comment by maracuja-juice** [comment]
I'm referring that it's generally not possible to achieve what I want to achieve in an accessible way.
 It feels like at this point I'm reiterating the same things again but I'll gladly do it because this is important to me. 

Yes I quoted that example already before. The menu items in the drawer are not direct siblings of the menu button (which btw in the markup is a link, which is also not entirely correct) and this makes it inaccessible to screen readers.

**Comment by abhiomkar** [comment]
I'm not sure what's the ARIA recommendation when menu button and menu items are not siblings. I noticed in many examples that menu popup is hoisted to body element instead of rendering as sibling to menu button.

Please update us with final example which is fully accessible. I'm sorry there is nothing we could do at this point.

**Comment by maracuja-juice** [comment]
There is no ARIA reccomendation for that. The **only** option is to have the items as siblings. (Just want to point out again that the sources I posted explain that very well.)

The markup I posted above (or similar) would be what I would suggest for future usage. 
The only important thing is that the items are adjacent to the hamburger button.

I don't understand why this was closed. Why is it not possible to put this on the backlog to change the library so that an accessible markup is possible for hamburger buttons?

**Comment by abhiomkar** [comment]
Proposed solution that you shared in this comment https://github.com/material-components/material-components-web/issues/4704#issuecomment-494434259 is not feasible for us to implement.

I feel moving DOM elements closer to hamberger that breaks the component layout is not appropriate solution to solve accessibility.

**Comment by maracuja-juice** [comment]
In the sources you can see that it is the only one that actually works. Because the aria-controls implementation in different screen readers is not good.

Unfortunately we won't be able to use this library now because it doesn't meet  the accessiblity standards.

**Comment by maracuja-juice** [comment]
Any update on why you don't want to make this library more accessible?
